[Gradient Compression] Add register_comm_hook API to DDP communication hooks documentation page

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51846 [Gradient Compression] Add register_comm_hook API to DDP communication hooks documentation page**

`register_comm_hook` method is defined in DistributedDataParallel module, but it is not covered in `distributed.rst`. Since it's closely related to DDP communication hook, add the docstrings to `ddp_comm_hooks.rst` instead of a reference.

Screenshot:
![Screen Shot 2021-02-06 at 4 17 10 PM](https://user-images.githubusercontent.com/8884619/107132819-1cf37600-6897-11eb-9643-f1cb42c4981d.png)


Differential Revision: [D26298191](https://our.internmc.facebook.com/intern/diff/D26298191/)

Commit:
[Gradient Compression] Add register_comm_hook API to DDP communication hooks documentation page

`register_comm_hook` method is defined in DistributedDataParallel module, but it is not covered in `distributed.rst`. Since it's closely related to DDP communication hook, add the docstrings to `ddp_comm_hooks.rst` instead of a reference.

Screenshot:

{F370425625}

Differential Revision: [D26298191](https://our.internmc.facebook.com/intern/diff/D26298191/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Add register_comm_hook API to DDP communication hooks documentation page"


`register_comm_hook` method is defined in DistributedDataParallel module, but it is not covered in `distributed.rst`. Since it's closely related to DDP communication hook, add the docstrings to `ddp_comm_hooks.rst` instead of a reference.

Screenshot:
![Screen Shot 2021-02-06 at 4 17 10 PM](https://user-images.githubusercontent.com/8884619/107132819-1cf37600-6897-11eb-9643-f1cb42c4981d.png)


Differential Revision: [D26298191](https://our.internmc.facebook.com/intern/diff/D26298191/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Add register_comm_hook API to DDP communication hooks documentation page"


`register_comm_hook` method is defined in DistributedDataParallel module, but it is not covered in `distributed.rst`. Since it's closely related to DDP communication hook, add the docstrings to `ddp_comm_hooks.rst` instead of a reference.

Screenshot:
![Screen Shot 2021-02-06 at 4 17 10 PM](https://user-images.githubusercontent.com/8884619/107132819-1cf37600-6897-11eb-9643-f1cb42c4981d.png)


Differential Revision: [D26298191](https://our.internmc.facebook.com/intern/diff/D26298191/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Add register_comm_hook API to DDP communication hooks documentation page"


`register_comm_hook` method is defined in DistributedDataParallel module, but it is not covered in `distributed.rst`. Since it's closely related to DDP communication hook, add the docstrings to `ddp_comm_hooks.rst` instead of a reference.

Screenshot:
![Screen Shot 2021-02-06 at 4 17 10 PM](https://user-images.githubusercontent.com/8884619/107132819-1cf37600-6897-11eb-9643-f1cb42c4981d.png)


Differential Revision: [D26298191](https://our.internmc.facebook.com/intern/diff/D26298191/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Add register_comm_hook API to DDP communication hooks documentation page"


`register_comm_hook` method is defined in DistributedDataParallel module, but it is not covered in `distributed.rst`. Since it's closely related to DDP communication hook, add the docstrings to `ddp_comm_hooks.rst` instead of a reference.

Screenshot:
![Screen Shot 2021-02-06 at 4 17 10 PM](https://user-images.githubusercontent.com/8884619/107132819-1cf37600-6897-11eb-9643-f1cb42c4981d.png)


Differential Revision: [D26298191](https://our.internmc.facebook.com/intern/diff/D26298191/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Add register_comm_hook API to DDP communication hooks documentation page"


`register_comm_hook` method is defined in DistributedDataParallel module, but it is not covered in `distributed.rst`. Since it's closely related to DDP communication hook, add the docstrings to `ddp_comm_hooks.rst` instead of a reference.

Screenshot:
![Screen Shot 2021-02-06 at 4 17 10 PM](https://user-images.githubusercontent.com/8884619/107132819-1cf37600-6897-11eb-9643-f1cb42c4981d.png)


Differential Revision: [D26298191](https://our.internmc.facebook.com/intern/diff/D26298191/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Add register_comm_hook API to DDP communication hooks documentation page"


`register_comm_hook` method is defined in DistributedDataParallel module, but it is not covered in `distributed.rst`. Since it's closely related to DDP communication hook, add the docstrings to `ddp_comm_hooks.rst` instead of a reference.

Screenshot:
![Screen Shot 2021-02-06 at 4 17 10 PM](https://user-images.githubusercontent.com/8884619/107132819-1cf37600-6897-11eb-9643-f1cb42c4981d.png)


Differential Revision: [D26298191](https://our.internmc.facebook.com/intern/diff/D26298191/)

[ghstack-poisoned]

rohan-varma reviewed (approved):
LGTM, although I think the doc build is failing, please take a look at that:
```
Feb 08 06:37:28 waiting for workers...
Feb 08 06:37:45 /opt/conda/lib/python3.6/site-packages/torch/distributed/distributed_c10d.py:144: UserWarning: torch.distributed.reduce_op is deprecated, please use torch.distributed.ReduceOp instead
Feb 08 06:37:45   warnings.warn("torch.distributed.reduce_op is deprecated, please use "
Feb 08 06:38:10 /var/lib/jenkins/workspace/docs/source/generated/torch.nn.parallel.DistributedDataParallel.rst:358: WARNING: duplicate object description of torch.nn.parallel.DistributedDataParallel.register_comm_hook, other instance in ddp_comm_hooks, use :noindex: for one of them
Feb 08 06:38:10 looking for now-outdated files... none found
Feb 08 06:38:10 pickling environment... done
Feb 08 06:38:10 checking consistency... done
Feb 08 06:38:10 running tests...
Feb 08 06:38:11 
```

Commit:
Update on "[Gradient Compression] Add register_comm_hook API to DDP communication hooks documentation page"


`register_comm_hook` method is defined in DistributedDataParallel module, but it is not covered in `distributed.rst`. Since it's closely related to DDP communication hook, add the docstrings to `ddp_comm_hooks.rst` instead of a reference.

Screenshot:
![Screen Shot 2021-02-06 at 4 17 10 PM](https://user-images.githubusercontent.com/8884619/107132819-1cf37600-6897-11eb-9643-f1cb42c4981d.png)


Differential Revision: [D26298191](https://our.internmc.facebook.com/intern/diff/D26298191/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Add register_comm_hook API to DDP communication hooks documentation page"


`register_comm_hook` method is defined in DistributedDataParallel module, but it is not covered in `distributed.rst`. Since it's closely related to DDP communication hook, add the docstrings to `ddp_comm_hooks.rst` instead of a reference.

Screenshot:
![Screen Shot 2021-02-06 at 4 17 10 PM](https://user-images.githubusercontent.com/8884619/107132819-1cf37600-6897-11eb-9643-f1cb42c4981d.png)


Differential Revision: [D26298191](https://our.internmc.facebook.com/intern/diff/D26298191/)

[ghstack-poisoned]
